migrate_cpp/rewriter.cpp

```diff
@@ -0,0 +1,284 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "migrate_cpp/rewriter.h"
+
+#include "clang/Tooling/Tooling.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/FormatVariadic.h"
+
+namespace Carbon {
+
+static constexpr const char CppPlaceholder[] = "__cpp__{ ... }";
+
+auto OutputWriter::Write(clang::SourceLocation loc,
+                         const OutputSegment& segment) const -> bool {
+  return std::visit(
+      [&](auto& content) {
+        using type = std::decay_t<decltype(content)>;
+        auto [begin, end] = bounds;
+
+        if constexpr (std::is_same_v<type, std::string>) {
+          auto begin_offset = source_manager.getDecomposedLoc(loc).second;
+          // Append the string replacement if the node being replaced falls
+          // within `bounds`.
+          if (begin <= begin_offset && begin_offset < end) {
+            output.append(content);
+          }
+        } else if constexpr (std::is_same_v<type, clang::DynTypedNode> ||
+                             std::is_same_v<type, clang::TypeLoc>) {
+          auto content_loc = content.getSourceRange().getBegin();
+          auto begin_offset =
+              source_manager.getDecomposedLoc(content_loc).second;
+          // If the node we're considering a replacement for is already beyond
+          // the region for which we want to make a replacement, exit early
+          // declaring that we have completed replacements (by returning false).
+          // Otherwise proceed. Note that we do not exit early or skip anything
+          // if the node comes before the relevant region. This is because many
+          // nodes in Clang's AST have a starting source location but a
+          // meaningless end location, and while the start of the segment may
+          // not be in the range, as we recurse, sub-segments may indeed end up
+          // being printed.
+          if (begin_offset >= end) {
+            return false;
+          }
+
+          if (auto iter = map.find(content); iter == map.end()) {
+            output.append(CppPlaceholder);
+          } else {
+            for (const auto& output_segment : iter->second) {
+              if (!Write(content.getSourceRange().getBegin(), output_segment)) {
+                return false;
+              }
+            }
+          }
+        } else {
+          static_assert(std::is_void_v<type>);
```

Comment:
I should ask -- is this ever called with void, or could this use CARBON_FATAL?

Reply:
This is never called with void (parameters can never be void). It ensures we've covered all cases. Changing this to CARBON_FATAL would reduce a compile-time failure to a run-time failure. I added a message that might help clarify a bit.